From 312a8f38dd6ccf3aafb354b7ad6a9a800669ab44 Mon Sep 17 00:00:00 2001
From: augustine saidimu <47994458+augustinesaidimu@users.noreply.github.com>
Date: Wed, 2 Feb 2022 03:57:53 +0300
Subject: [PATCH] chore: applied node theme

---
 config.def.h | 53 ++++++++++++++++++++++++++--------------------------
 1 file changed, 26 insertions(+), 27 deletions(-)

diff --git a/config.def.h b/config.def.h
index 91ab8ca..658e620 100644
--- a/config.def.h
+++ b/config.def.h
@@ -4,8 +4,9 @@
  * appearance
  *
  * font: see http://freedesktop.org/software/fontconfig/fontconfig-user.html
+ * Nerd Fonts: see https://www.nerdfonts.com/
  */
-static char *font = "Liberation Mono:pixelsize=12:antialias=true:autohint=true";
+static char *font = "JetBrainsMono Nerd Font Mono:pixelsize=14:antialias=true:autohint=true";
 static int borderpx = 2;

 /*
@@ -16,7 +17,7 @@ static int borderpx = 2;
  * 4: value of shell in /etc/passwd
  * 5: value of shell in config.h
  */
-static char *shell = "/bin/sh";
+static char *shell = "/bin/zsh";
 char *utmp = NULL;
 /* scroll program: to enable use a string like "scroll" */
 char *scroll = NULL;
@@ -96,32 +97,30 @@ unsigned int tabspaces = 8;
 /* Terminal colors (16 first used in escape sequence) */
 static const char *colorname[] = {
 	/* 8 normal colors */
-	"black",
-	"red3",
-	"green3",
-	"yellow3",
-	"blue2",
-	"magenta3",
-	"cyan3",
-	"gray90",
+   "#1a181a", /* black   */
+   "#f85e84", /* red     */
+   "#9ecd6f", /* green   */
+   "#e5c463", /* yellow  */
+   "#7accd7", /* blue    */
+   "#ab9df2", /* magenta */
+   "#ef9062", /* cyan    */
+   "#e3e1e4", /* white   */

 	/* 8 bright colors */
-	"gray50",
-	"red",
-	"green",
-	"yellow",
-	"#5c5cff",
-	"magenta",
-	"cyan",
-	"white",
+   "#848089", /* black   */
+   "#f85e84", /* red     */
+   "#9ecd6f", /* green   */
+   "#e5c463", /* yellow  */
+   "#7accd7", /* blue    */
+   "#ab9df2", /* magenta */
+   "#ef9062", /* cyan    */
+   "#e3e1e4", /* white   */

 	[255] = 0,

 	/* more colors can be added after 255 to use with DefaultXX */
-	"#cccccc",
-	"#555555",
-	"gray90", /* default foreground colour */
-	"black", /* default background colour */
+   "#2d2a2e", /* background */
+   "#e3e1e4", /* foreground */
 };


@@ -129,10 +128,10 @@ static const char *colorname[] = {
  * Default colors (colorname index)
  * foreground, background, cursor, reverse cursor
  */
-unsigned int defaultfg = 258;
-unsigned int defaultbg = 259;
-unsigned int defaultcs = 256;
-static unsigned int defaultrcs = 257;
+unsigned int defaultfg = 257;
+unsigned int defaultbg = 256;
+unsigned int defaultcs = 257;
+static unsigned int defaultrcs = 256;

 /*
  * Default shape of cursor
@@ -141,7 +140,7 @@ static unsigned int defaultrcs = 257;
  * 6: Bar ("|")
  * 7: Snowman ("☃")
  */
-static unsigned int cursorshape = 2;
+static unsigned int cursorshape = 4;

 /*
  * Default columns and rows numbers
--
2.34.1

